File: src/MaintenanceSystem.Domain/Entities/ChecklistStep.cs
namespace MaintenanceSystem.Domain.Entities;

public class ChecklistStep
{
    public Guid Id { get; private set; }
    public Guid WorkOrderId { get; private set; }
    public string Label { get; private set; } = string.Empty;
    public bool IsMandatory { get; private set; }
    public bool IsCompleted { get; private set; }
    public DateTime? CompletedAt { get; private set; }

    private ChecklistStep() { }

    public static ChecklistStep Create(Guid workOrderId, string label, bool isMandatory) => new()
    {
        Id = Guid.NewGuid(),
        WorkOrderId = workOrderId,
        Label = label,
        IsMandatory = isMandatory,
        IsCompleted = false
    };

    public void Complete()
    {
        IsCompleted = true;
        CompletedAt = DateTime.UtcNow;
    }

    public void Uncomplete()
    {
        IsCompleted = false;
        CompletedAt = null;
    }
}

File: app/lib/models/checklist_step.dart
class ChecklistStep {
  final String id;
  final String workOrderId;
  final String label;
  final bool isMandatory;
  final bool requiresPhoto;
  final bool isCompleted;
  final String? completedAt;

  const ChecklistStep({
    required this.id,
    required this.workOrderId,
    required this.label,
    required this.isMandatory,
    required this.requiresPhoto,
    required this.isCompleted,
    this.completedAt,
  });

  factory ChecklistStep.fromJson(Map<String, dynamic> j) => ChecklistStep(
        id: j['id'] as String,
        workOrderId: j['workOrderId'] as String,
        label: j['label'] as String,
        isMandatory: j['isMandatory'] as bool,
        requiresPhoto: j['requiresPhoto'] as bool? ?? false,
        isCompleted: j['isCompleted'] as bool,
        completedAt: j['completedAt'] as String?,
      );

  ChecklistStep copyWith({bool? isCompleted}) => ChecklistStep(
        id: id,
        workOrderId: workOrderId,
        label: label,
        isMandatory: isMandatory,
        requiresPhoto: requiresPhoto,
        isCompleted: isCompleted ?? this.isCompleted,
        completedAt: completedAt,
      );
}
